### ISSUE TITLE
unsetting creations, updates... in time

### ISSUE THREAD

**Issue by toooni** [w3c/W3CLifecycleEventsBundle#8]
https://github.com/w3c/W3CLifecycleEventsBundle/blob/4d894b38d21af12f8242d9385433a3c7d03d75fe/Services/LifecycleEventsDispatcher.php#L118-L128

Before preparing a PR, I wanted to ask if there is a reason the `$this->creations` array is only "unsetted" at the end?
I've got the problem that a `persist` and `flush` inside the listener with the same entity does it trigger twice because it isn't unsetted in time.

My proposition would be like this:
```
     private function dispatchCreationEvents() 
     { 
         foreach ($this->creations as $key => $creation) { 
             $annotation = $creation[0]; 
             /** @var LifecycleEventArgs $eventArgs */ 
             $eventArgs  = $creation[1]; 
             $entity     = $eventArgs->getObject(); 
  
             unset($this->creations[$key]);
             $this->dispatcher->dispatch($annotation->event, new $annotation->class($entity)); 
         } 
     } 
```

**Comment by jean-gui** [comment]
I haven't worked on this bundle in a while, but if I get this right, performing a `flush` in the listener does indeed trigger another dispatchEvents, which will then dispatch all events a second time because the five arrays are only cleared at the end of dispatchEvents.

Regarding your proposed change, I've never liked unsetting things that you are currently iterating on, but more importantly I don't think that would fix the issue: all elements that have been processed wouldn't be processed again, but those that haven't would still be processed twice. 

The code would probably have to be smarter. Maybe when we start dispatching events, we should set a bit that would lock the arrays, prevent another dispatchEvents from running and store new events to secondary arrays. At the end of dispatchEvents, move those new events back to the main arrays, and recursively call dispatchEvents again. A protection would be necessary to prevent infinite recursion.

How does that sound?

**Comment by toooni** [comment]
Yeah, that‘s right.

I am not completely sure but it‘s possible you‘re right about the not processed events.

IMO this should be a solution in any cases:
```
private function dispatchCreationEvents() 
     { 
         $creations = $this->creations;
         $this->creations = [];

         foreach ($creations as $key => $creation) { 
             $annotation = $creation[0]; 
             /** @var LifecycleEventArgs $eventArgs */ 
             $eventArgs  = $creation[1]; 
             $entity     = $eventArgs->getObject(); 
  
             $this->dispatcher->dispatch($annotation->event, new $annotation->class($entity)); 
         } 
     } 
```

**Comment by toooni** [comment]
Sorry, weote my command on my smartphone. Can‘t fix the formatting yet ;)

**Comment by jean-gui** [comment]
Doh, that's much simpler than my convoluted idea. Fee free to submit a PR.

**Comment by jean-gui** [comment]
Fixed in v2.0.0
